Mathlib/Algebra/Homology/ComplexShape.lean

```diff
@@ -0,0 +1,198 @@
+/-
+Copyright (c) 2021 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Scott Morrison
+-/
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Logic.Relation
+
+/-!
+# Shapes of homological complexes
+
+We define a structure `ComplexShape ι` for describing the shapes of homological complexes
+indexed by a type `ι`.
+This is intended to capture chain complexes and cochain complexes, indexed by either `ℕ` or `ℤ`,
+as well as more exotic examples.
+
+Rather than insisting that the indexing type has a `succ` function
+specifying where differentials should go,
+inside `c : ComplexShape` we have `c.rel : ι → ι → Prop`,
+and when we define `HomologicalComplex`
+we only allow nonzero differentials `d i j` from `i` to `j` if `c.rel i j`.
+Further, we require that `{ j // c.rel i j }` and `{ i // c.rel i j }` are subsingletons.
+This means that the shape consists of some union of lines, rays, intervals, and circles.
+
+Convenience functions `c.next` and `c.prev` provide these related elements
+when they exist, and return their input otherwise.
+
+This design aims to avoid certain problems arising from dependent type theory.
+In particular we never have to ensure morphisms `d i : X i ⟶ X (succ i)` compose as
+expected (which would often require rewriting by equations in the indexing type).
+Instead such identities become separate proof obligations when verifying that a
+complex we've constructed is of the desired shape.
+
+If `α` is an `AddRightCancelSemigroup`, then we define `up α : ComplexShape α`,
+the shape appropriate for cohomology,so `d : X i ⟶ X j` is nonzero only when `j = i + 1`,
+as well as `down α : ComplexShape α`, appropriate for homology,
+so `d : X i ⟶ X j` is nonzero only when `i = j + 1`.
+(Later we'll introduce `CochainComplex` and `ChainComplex` as abbreviations for
+`HomologicalComplex` with one of these shapes baked in.)
+-/
+
+open Classical
+
+noncomputable section
+
+/-- A `c : ComplexShape ι` describes the shape of a chain complex,
+with chain groups indexed by `ι`.
+Typically `ι` will be `ℕ`, `ℤ`, or `fin n`.
+
+There is a relation `rel : ι → ι → Prop`,
+and we will only allow a non-zero differential from `i` to `j` when `rel i j`.
+
+There are axioms which imply `{ j // c.rel i j }` and `{ i // c.rel i j }` are subsingletons.
+This means that the shape consists of some union of lines, rays, intervals, and circles.
+
+Below we define `c.next` and `c.prev` which provide these related elements.
+-/
+@[ext]
+structure ComplexShape (ι : Type _) where
+  rel : ι → ι → Prop
+  next_eq : ∀ {i j j'}, rel i j → rel i j' → j = j'
+  prev_eq : ∀ {i i' j}, rel i j → rel i' j → i = i'
+#align complex_shape ComplexShape
```

Comment:
If you decide no name change is necessary beyond what mathport suggested, it's allowed (and possibly preferred??) to delete the #align.

Reply:
From the mathport output, the only name change I did was for the field `rel` (which mathport renamed `Rel`). Then, should I remove all the `#align` in this file?

Reply:
the field should be called `Rel`.

Reply:
(Because its conclusion is a `Prop` or `Type`.)

Reply:
Thanks!